contracts/FeralfileArtworkV4_2.sol

```diff
@@ -0,0 +1,184 @@
+
+            if (!_exists(tokenIdForSale)) {
+                revert TokenIDNotFound();
+            }
+
+            nextSaleTokenId++;
+            if (ownerOf(tokenIdForSale) != address(this)) {
+                continue;
+            }
+
+            // send NFT
+            _safeTransfer(
+                address(this),
+                saleData_.destination,
+                tokenIdForSale,
+                ""
+            );
+
+            emit BuyArtworkV2(saleData_.destination, tokenIdForSale, saleData_.nonce);
+            i++;
+        }
+
+        // save next sale token id for seriesID
+        seriesNextSaleTokenIds[saleData_.seriesID] = nextSaleTokenId;
+
+        // distribute royalty
+        uint256 distributedRevenue;
+        uint256 platformRevenue;
+
+        RevenueShare[] memory revenueShares = saleData_.revenueShares;
+        uint256 remainingRev = itemRevenue;
+
+        // deduct advances payment from revenue
+        for (uint256 j = 0; j < revenueShares.length && remainingRev > 0; j++) {
+            uint256 remainingAdvanceAmount = advances[
+                revenueShares[j].recipient
+            ];
+            uint256 rev = remainingAdvanceAmount >= remainingRev
+                ? remainingRev
+                : remainingAdvanceAmount;
+            uint256 totalRev = saleData_.quantity * rev;
+            platformRevenue += totalRev;
+            advances[revenueShares[j].recipient] -= totalRev;
+            remainingRev -= totalRev;
+        }
+
+        // distribute revenue
+        if (remainingRev > 0) {
+            for (uint256 j = 0; j < revenueShares.length; j++) {
+                address recipient = revenueShares[j].recipient;
+                uint256 rev = (remainingRev * revenueShares[j].bps) / 10000;
+                uint256 totalRev = saleData_.quantity * rev;
+                if (recipient == costReceiver) {
+                    platformRevenue += totalRev;
+                    continue;
+                }
+                distributedRevenue += totalRev;
+                payable(recipient).transfer(totalRev);
+            }
+        }
+
+        require(
+            saleData_.price - saleData_.cost >=
+                distributedRevenue + platformRevenue,
+            "FeralfileExhibitionV4: total bps over 10,000"
+        );
+
+        // Transfer cost, platform revenue and remaining funds
+        uint256 leftOver = saleData_.price - distributedRevenue;
+        if (leftOver > 0) {
+            payable(costReceiver).transfer(leftOver);
+        }
+    }
+
+    /// @notice override revert buyArtworks
+    function buyArtworks(
+        bytes32,
+        bytes32,
+        uint8,
+        SaleData calldata
+    ) external payable override {
+        revert FunctionNotSupported();
+    }
+
+    /// @notice Event emitted when Artwork has been sold with the additional nonce
+    event BuyArtworkV2(address indexed buyer, uint256 indexed tokenId, uint256 nonce);
```

Comment:
Should be `buyBulkArtworks`

Reply:
currently, the event is for each single token

Reply:
@lpopo0856 What would you prefer? Single event for this bulk sale or multiple events for individual?